src/interface/state_descriptor.hpp

```diff
@@ -412,31 +413,41 @@ class StateDescriptor {
     if (UserWorkBeforeLoopMesh != nullptr) return UserWorkBeforeLoopMesh(pmesh, pin, tm);
   }
 
+  void UserWorkBeforeOutput(Mesh *pmesh, ParameterInput *pin, SimTime &tm) const {
+    if (UserWorkBeforeOutputMesh) return UserWorkBeforeOutputMesh(pmesh, pin, tm);
+  }
+
+  void UserWorkBeforeRestartOutput(Mesh *pmesh, ParameterInput *pin, SimTime &tm,
+                                   OutputParameters *pparams) const {
+    if (UserWorkBeforeRestartOutputMesh)
+      return UserWorkBeforeRestartOutputMesh(pmesh, pin, tm, pparams);
+  }
+
   std::vector<std::shared_ptr<AMRCriteria>> amr_criteria;
 
-  std::function<void(MeshBlockData<Real> *rc)> PreCommFillDerivedBlock = nullptr;
-  std::function<void(MeshData<Real> *rc)> PreCommFillDerivedMesh = nullptr;
-  std::function<void(MeshBlockData<Real> *rc)> PreFillDerivedBlock = nullptr;
-  std::function<void(MeshData<Real> *rc)> PreFillDerivedMesh = nullptr;
-  std::function<void(MeshBlockData<Real> *rc)> PostFillDerivedBlock = nullptr;
-  std::function<void(MeshData<Real> *rc)> PostFillDerivedMesh = nullptr;
-  std::function<void(MeshBlockData<Real> *rc)> FillDerivedBlock = nullptr;
-  std::function<void(MeshData<Real> *rc)> FillDerivedMesh = nullptr;
-  std::function<void(Mesh *, ParameterInput *, SimTime &)> UserWorkBeforeLoopMesh =
-      nullptr;
-
-  std::function<void(SimTime const &simtime, MeshData<Real> *rc)> PreStepDiagnosticsMesh =
-      nullptr;
-  std::function<void(SimTime const &simtime, MeshData<Real> *rc)>
-      PostStepDiagnosticsMesh = nullptr;
-
-  std::function<Real(MeshBlockData<Real> *rc)> EstimateTimestepBlock = nullptr;
-  std::function<Real(MeshData<Real> *rc)> EstimateTimestepMesh = nullptr;
-
-  std::function<AmrTag(MeshBlockData<Real> *rc)> CheckRefinementBlock = nullptr;
-
-  std::function<void(MeshData<Real> *rc)> InitNewlyAllocatedVarsMesh = nullptr;
-  std::function<void(MeshBlockData<Real> *rc)> InitNewlyAllocatedVarsBlock = nullptr;
+  std::function<void(MeshBlockData<Real> *rc)> PreCommFillDerivedBlock;
+  std::function<void(MeshData<Real> *rc)> PreCommFillDerivedMesh;
+  std::function<void(MeshBlockData<Real> *rc)> PreFillDerivedBlock;
+  std::function<void(MeshData<Real> *rc)> PreFillDerivedMesh;
+  std::function<void(MeshBlockData<Real> *rc)> PostFillDerivedBlock;
+  std::function<void(MeshData<Real> *rc)> PostFillDerivedMesh;
+  std::function<void(MeshBlockData<Real> *rc)> FillDerivedBlock;
+  std::function<void(MeshData<Real> *rc)> FillDerivedMesh;
+  std::function<void(Mesh *, ParameterInput *, SimTime &)> UserWorkBeforeLoopMesh;
+  std::function<void(Mesh *, ParameterInput *, SimTime &)> UserWorkBeforeOutputMesh;
```

Comment:
The `= nullptr` and `= {nullptr}` are redundant so I removed them.

Reply:
C++ trivia for the day. Apparently `std::function` very explicitly has an overload of `operator==` for comparing against `nullptr_t`. The zero-initialized (and initialized to nullptr) `std::function` is actually an *empty* `std::function`.

Reply:
https://en.cppreference.com/w/cpp/utility/functional/function/operator_cmp

Reply:
So we traded the `= nullptr` against `if (FunctionCall) {FunctionCall()}`?
Are there any C++ best-practices for one or the other?
To me the original approach looked cleaner (though just commenting -- not saying we should revert).

Reply:
Yes I don't think we traded any behavior for how we check whether these function pointers are initialized or not -- comparisons with `std::function` behave the same whether we explicitly initialize to `nullptr` or not. 

I went back and forth on this and decided the self-documenting quality of `= nullptr` and `!= nullptr` is valuable so I reverted to what we've done previously. 

I couldn't find any style guidelines about this in e.g. the google style guide.

Reply:
I was actually referring to other changes (see following highlight)